### DISCUSSION TITLE
Implementing New PMD Rule

### DISCUSSION THREAD

**Discussion by ShreyChandra** [pmd/pmd#3466] [Q&A] [answered]
Hi,
I am trying to implement a rule where I need to check an if statement's condition with the next else if's condition, something like:

```
public class Main {
  public static void main(String[] args) {
    int time = 22;
    if (condition A) {
      System.out.println("Good morning.");
    } else if (condition not A) { 
      System.out.println("Good day.");
    }
  }
}
```
So I want to extract the two consecutive conditions and then compare them (string comparison).

I was able to extract the two conditions using XPATH separately but was super confused about how I should compare the two extracted conditions as two separate strings.

The other way was to use Java to write the rule but then I couldn't find an existing rule from which I could learn and adapt, mostly how to extract the conditions from the if and else if statements. If I can get them as strings it would be quite easy to compare them.

The only way I could think of tackling this problem was that if I could use Xpath to extract the conditions and then compare them using Java. Although am not sure if is valid or even possible?

If someone could guide me a bit as to how to approach this problem, I'd be really grateful. Thanks !

**Comment by oowekyala** [comment, marked as answer]
So to decompose the problem your rule tries to solve, you want to
- find an `if` statement that has an `else if` part
- compare both conditions to find if the second is the negation of the first

The second part is impossible to do in XPath, so I'll just go ahead with Java.
Using a Java rule you can implement a [visit method](https://pmd.github.io/latest/pmd_userdocs_extending_writing_java_rules.html#tree-traversal) to visit all `if` statements:

```java
    @Override
    public Object visit(ASTIfStatement fstIf, Object data) {
        ASTIfStatement sndIf = getAsIfStmt(fstIf.getElseBranch());
        if (sndIf != null) {// it's an if/else if
            if (isNegation(fstIf.getCondition(), sndIf.getCondition())) {
                addViolation(data, sndIf);
            }
        }
        return data;
    }

    private ASTIfStatement getAsIfStmt(ASTStatement stmt) {
        if (stmt != null && stmt.getChild(0) instanceof ASTIfStatement) {
            return (ASTIfStatement) stmt.getChild(0);
        }
        return null;
    }

    /** Returns true if the first expression is the negation of the second. */
    private boolean isNegation(ASTExpression e1, ASTExpression e2) {
        // todo
    }
```

Then the remaining problem is to implement `isNegation`. This is not trivial if you want to handle eg de morgan laws and operator symmetry and such. But let's just implement the specific case that recognizes that the negation of `==` is  `!=`. Intuitively, the body of your `isNegation` will look like so:
```java
if (e1 instanceof ASTEqualityExpression && e2 instanceof ASTEqualityExpression) {
   // e1 is `e11 == e12` or `e11 != e12`
   // e2 is `e21 == e22` or `e21 != e22`
   JavaNode e11 = e1.getChild(0);
   JavaNode e12 = e1.getChild(1);
   JavaNode e21 = e2.getChild(0);
   JavaNode e22 = e2.getChild(1);

   // Given a function `eq` that tests equivalence of nodes,
   // e1 is the negation of e2 if
   //   eq(e11, e21) && eq(e12, e22)  // ie, e1 and e2 have the same subexpressions (note, this test doesn't handle symmetry of ==)
   //   && !e1.getOperator().equals(e2.getOperator()) // either e1 is == and e2 is !=, or the reverse is true

}
return false;
```
This assumes we have an equivalence function `eq`. You may just test equality token by token. Getting expressions as strings is not supported, and it's also a non-goal of ours, as it's more powerful to use tokens / AST.

Note that this has been implemented in a more complete way in the pmd 7 development branch. It takes advantages of simplifications to the grammar and some new apis, so it cannot be backported as is to pmd 6. But you could use it as inspiration (including the token comparison routine). See 

https://github.com/pmd/pmd/blob/b0b20873162bda29c6c048a363b5897eb8bdb3e7/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/JavaRuleUtil.java#L565-L656

**Reply by ShreyChandra** [reply]
Thank you so much for the detailed explanation. :)
